### PR TITLE
Adds Standard transformer flops formula to flops callback

### PR DESCRIPTION
**Pull request by jomitchellnv**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds the Transformer flops calculation from here https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/main/megatron/training/training.py?ref_type=heads#L199 into the NeMo FlopsFormulas.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Adds a Flops Formula for generic Transformer models

# Usage

- You can use this by adding a flops callback such as 
```python
flops_callback = FLOPsMeasurementCallback(
        model_config=bionemo_model_config,
        data_config=data_module,
        model_name="transformer",
    )
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/flops_callback.py [hunk 1]
```diff
@@ -132,6 +132,14 @@ def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx: int)
                 batch_size=1,
                 prog_bar=True,
             )
+            pl_module.log(
+                "total_flops",
```
[the remark this answers is not in the document]
Comment (dorotat-nv) [reply nemo/lightning/pytorch/callbacks/flops_callback.py:136]: this would be a very handy method for MFU calculation

**Review by dorotat-nv — CHANGES_REQUESTED** [review]
I suggest passing the `vocab_size` mapping directly in BioNeMo and using config.vocab_size consistently. This will help avoid confusion, especially when any vocabulary-specific parameters change. Hence removing `geneformer` mapping and ensuring that `vocab_size` is explicitly passed within BioNeMo.

Additionally, I propose logging model_tflops_per_gpu instead of total_flops, as the latter is unclear (total flops of what, exactly?).

Lastly, is the transformer method intended to apply to all transformer models? How do the other calculation methods relate to or differ from the transformer method?

**Comment by jomitchellnv** [comment]
@NVIDIA/nemo-core-reviewers

File: nemo/collections/common/parts/perf_metrics_utils.py [hunk 2]
```diff
@@ -26,6 +26,7 @@
     "nemotron": 256000,
     "bert": 29000,
     "mixtral": 32000,
+    "geneformer": 25429,
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply nemo/collections/common/parts/perf_metrics_utils.py:29]: ok ive removed it

File: nemo/lightning/pytorch/callbacks/flops_callback.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply nemo/lightning/pytorch/callbacks/flops_callback.py:136]: Done

**Comment by deepakn94** [comment]
LGTM.

**Review by dorotat-nv — CHANGES_REQUESTED** [review]
Found one bug, otherwise, lgtm

File: nemo/utils/flops_formulas.py [hunk 3]
```diff
@@ -150,6 +151,103 @@ def bert(config: FLOPSConfig):
...
+        # This can be expanded based on NeMo's actual MoE implementation
+        num_moe_layers = num_layers // 2  # Simplified assumption
+        num_dense_layers = num_layers - num_moe_layers
+        num_experts_routed_to = moe_router_topk
+
+    # Handle SwiGLU vs standard GELU/ReLU
+    # Default to standard activation (no SwiGLU)
+    gated_linear_multiplier = 1
+
+    # Define the expansion factor as described in the paper
```
Comment (malay-nagda) [inline nemo/utils/flops_formulas.py:192]: nitpick: can you add link to the paper in this comment?

File: nemo/utils/flops_formulas.py [hunk 4]
```diff
@@ -150,6 +151,103 @@ def bert(config: FLOPSConfig):
...
+def transformer(config: FLOPSConfig):
+    """Calculate FLOPs for a standard Transformer model."""
+    # Extract parameters from config
+    batch_size = config.gbs
+    hidden_size = config.hs
+    seq_length = config.enc_seq_len
+    num_layers = config.layers
+    num_attention_heads = config.attention_heads
+    ffn_hidden_size = config.ffn_hs
+    vocab_size = config.vocab_size or LLM_VOCAB_SIZE_MAP["geneformer"]
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply nemo/utils/flops_formulas.py]: Fixed

File: nemo/utils/flops_formulas.py (hunk 3 above)
Comment (jomitchellnv) [reply nemo/utils/flops_formulas.py:192]: I don't have the paper. I took this formula from here: https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/main/megatron/training/training.py?ref_type=heads#L194 
The previous formula for `bert_flops`... hardcodes ffn_h as 4*h and 
Over-counts the number of s*s floating-point operations by a factor of 2 for causal models where you only need to materialize half of the attention scores.

File: nemo/lightning/pytorch/callbacks/flops_callback.py [hunk 5]
```diff
@@ -132,6 +133,14 @@ def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx: int)
...
                 prog_bar=True,
             )
+            pl_module.log(
+                "total_model_tflops",
+                total_model_tflops,
+                on_step=True,
+                on_epoch=False,
+                batch_size=1,
+                prog_bar=True,
+            )
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply nemo/lightning/pytorch/callbacks/flops_callback.py]: Done. I added some logic to achieve this. Please double check if you like it.

File: nemo/utils/flops_formulas.py (hunk 3 above)
Comment (deepakn94) [reply nemo/utils/flops_formulas.py:192]: You can use this paper link for the comment: https://arxiv.org/pdf/2104.04473.

**Comment by jomitchellnv** [comment]
I suppose when I clicked update branch this entire PR fell apart and now it has a ton of code in it that I did not intend to be there.
